src/main/java/com/google/sps/servlets/PreviousTerms.java

```diff
@@ -0,0 +1,112 @@
+    String prevTermsDataJSON = makeJSON(prevTermsData);
+    response.setContentType("application/json;");
+    response.getWriter().println(prevTermsDataJSON);
+  }
+
+  private List<Entity> getPreviousTerms(DatastoreService db, HttpServletRequest request) {
+    String schoolName = request.getParameter("school-name");
+    String courseName = request.getParameter("course-name");
+    String termName = request.getParameter("term");
+    String profName = request.getParameter("prof-name");
+    Long units = Long.parseLong(request.getParameter("num-units"));
+
+    Entity foundTerm = findTerm(db, request);
+    Date startTime = (Date) foundTerm.getProperty("timeStamp");
+    Key schoolKey = findQueryMatch(db, "School", "school-name", schoolName).get(0).getKey();
+
+    List<Filter> filters = new ArrayList();
+    Filter courseFilter = new FilterPredicate("course-name", FilterOperator.EQUAL, courseName);
+    Filter unitFilter = new FilterPredicate("units", FilterOperator.EQUAL, units);
+    filters.add(courseFilter);
+    filters.add(unitFilter);
+
+    Query courseQuery =
+        new Query("Course").setAncestor(schoolKey).setFilter(CompositeFilterOperator.and(filters));
+    Key courseKey =
+        db.prepare(courseQuery).asList(FetchOptions.Builder.withDefaults()).get(0).getKey();
+
+    Filter timeFilter = new FilterPredicate("timeStamp", FilterOperator.LESS_THAN, startTime);
+    Query termQuery =
+        new Query("Term")
+            .setAncestor(courseKey)
+            .addSort("timeStamp", SortDirection.DESCENDING)
+            .setFilter(timeFilter);
+    List<Entity> foundTerms = db.prepare(termQuery).asList(FetchOptions.Builder.withLimit(2));
+    return foundTerms;
+  }
+
+  private Entity findTerm(DatastoreService db, HttpServletRequest request) {
+    String schoolName = request.getParameter("school-name");
+    String courseName = request.getParameter("course-name");
+    String termName = request.getParameter("term");
+    Long units = Long.parseLong(request.getParameter("num-units"));
+
+    Key schoolKey = findQueryMatch(db, "School", "school-name", schoolName).get(0).getKey();
+
+    List<Filter> filters = new ArrayList();
+    Filter courseFilter = new FilterPredicate("course-name", FilterOperator.EQUAL, courseName);
+    Filter unitFilter = new FilterPredicate("units", FilterOperator.EQUAL, units);
+    filters.add(courseFilter);
+    filters.add(unitFilter);
+
+    Query courseQuery =
+        new Query("Course").setAncestor(schoolKey).setFilter(CompositeFilterOperator.and(filters));
+    Key courseKey =
+        db.prepare(courseQuery).asList(FetchOptions.Builder.withDefaults()).get(0).getKey();
+
+    Filter termFilter = new FilterPredicate("term", FilterOperator.EQUAL, termName);
+    Query termQuery = new Query("Term").setAncestor(courseKey).setFilter(termFilter);
+    Entity foundTerm = db.prepare(termQuery).asList(FetchOptions.Builder.withDefaults()).get(0);
+
+    return foundTerm;
```

Comment:
One other quick question. In my PR you commented  my functions needing to return NOT_FOUND if a query comes back empty, but we are guaranteed that this can't be the case because these functions are called from a course page and we can only reach that course page if their corresponding entities exists in the DB. If this is the case should I still have a NOT_FOUND case or just leave a comment?

Reply:
Added a comment for the time being. When we shift to key/id in the url I will change it to throw a Datastore Note found error.